sherpa_contrib/diag_resp.py

```diff
@@ -0,0 +1,712 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301 USA.
+#
+
+
+
+
+""" Create diagonal RMF and flat ARF objects of value unity.
+
+Helper routines for creating a diagonal RMF and flat ARF response
+objects for various telescopes and instruments.
+The default is to generate Chandra/ACIS responses for PI-type spectra
+extracted from ACIS stowed background event files.
+
+Examples
+--------
+
+>>> from sherpa_contrib.diag_resp import *
+>>> diagrmf,flatarf = mkdiagresp()
+>>>
+>>> set_rmf(diagrmf, bkg_id=1)
+>>> set_arf(flatarf, bkg_id=1)
+
+or
+
+>>> diagrmf,flatarf = mkdiagresp(refspec="bkg.pi")
+
+or
+
+>>> bkgspec = get_bkg()
+>>> diagrmf,flatarf = mkdiagresp(refspec=bkgspec)
+
+"""
+
+__revision__ = "20 June 2024"
+
+import os
+import warnings
+from logging import getLogger
+from functools import wraps
+
+import numpy.typing as npt
+from numpy import arange
+from ciao_contrib._tools.fileio import get_keys_from_file
+
+from sherpa.astro.data import DataPHA, Data1DInt
+from sherpa.astro.ui import create_rmf as shpmkrmf
+from sherpa.astro.ui import create_arf as shpmkarf
+
+
+
+def _reformat_wmsg(func):
+    """
+    Decorator:
+
+    Trim extraneous user-warning information showing path and line number
+    where Sherpa warning message is being thrown for replacing 0s with ethresh.
+    """
+
+    @wraps(func)
+    def run_func(*args, **kwargs):
+        warnings_reset = warnings.formatwarning
+        warnings.formatwarning = lambda msg, *args_warn, **kwargs_warn: f'Warning: {msg}\n'
+
+        func_out = func(*args,**kwargs)
+
+        warnings.formatwarning = warnings_reset
+
+        return func_out
+
+    return run_func
+
+
+
+def _quash_shpverb(func):
+    """
+    Decorator:
+
+    Temporarily quash Sherpa verbosity.
+    """
+
+    def run_func(*args, **kwargs):
+        sherpalog = getLogger("sherpa")
+        loglevel = sherpalog.level
+        sherpalog.setLevel(0)
+
+        func_out = func(*args,**kwargs)
+
+        sherpalog.setLevel(loglevel)
+
+        return func_out
+
+    return run_func
+
+
+
+def _get_random_string(strlen: int = 16) -> str:
+    from string import ascii_letters,digits
+    from random import choices
+
+    chars = ascii_letters + digits
+
+    return ''.join(choices(chars, k=strlen))
+
+
+
+@_quash_shpverb
+def _get_file_header(fn) -> dict:
+    from sherpa.astro.ui import load_pha,get_data,delete_data
+
+    tmp_id = _get_random_string(strlen=32)
+
+    load_pha(tmp_id,fn)
+    hdr = get_data(tmp_id).header
+    delete_data(tmp_id)
+
+    return hdr
+
+
+
+class EGrid:
+    """
+    gather together all information for non-Chandra setups and return the energy-grid.
+    """
+
+    def __init__(self, telescope, instrument, detector, instfilter, nchan, chantype):
+        self.telescope = telescope
```

Comment:
I'd suggest you go ahead and `.lower()` and `.upper()` the values once in the constructor instead of each time you go to use them

Reply:
addressed in commit bad862a80b9d2a79a7c5aa7718962c30d28dc5af